src/betree.c

```diff
@@ -0,0 +1,1147 @@
+#include <math.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <ctype.h>
+
+#include "ast.h"
+#include "betree.h"
+
+bool match_sub(const struct event* event, const struct sub *sub)
+{
+    if(sub == NULL) {
+        return false;
+    }
+    return match_node(event, sub->expr) == 1;
+} 
+
+void check_sub(const struct event* event, const struct lnode* lnode, struct matched_subs* matched_subs) 
+{
+    for(unsigned int i=0; i < lnode->sub_count; i++) {
+        const struct sub* sub = lnode->subs[i];
+        if(match_sub(event, sub) == true) {
+            if(matched_subs->sub_count == 0) {
+                matched_subs->subs = malloc(sizeof(int));
+                if(matched_subs->subs == NULL) {
+                    fprintf(stderr, "check_sub malloc failed");
+                    exit(1);
```

Comment:
I'd recommend using `abort()` as it makes it much easier to debug with gdb as it can trap `SIGABRT` signal. You can still put a breakpoint on `exit` but I've always found `abort` to be more convenient.